File: matlab/example2.m
clear
clc

rng(100)

%% Load data
load('synthetic.mat')

for s = 1:length(synthetic)
    %% Set the number of clusters and fuzziness and compute the distance matrix
    k = synthetic{s}.k;
    h = 1.5;
    g = 1.5;
    d = pdist2(synthetic{s}.x,synthetic{s}.x).^2;
    
    %% Solve FKM, FMMdd and CFKM
    [Z_fkm,e_fkm] = FKM(d,k,h);
    [Z_fmmdd,e_fmmdd,v_fmmdd] = FMMdd(d,k,h,g);
    [Z_cfkm,e_cfkm] = CFKM(d,k,h);
    
    %% Plot FKM, FMMdd and CFKM solutions
    % Compute the PCA-projected points in two dimensions
    if size(synthetic{s}.x,2)==2
        x = synthetic{s}.x;
    else
        [~,x] = pca(synthetic{s}.x);
    end
    
    figure(s)
    
    % Plot actual points and clusters
    subplot(2,2,1)
    colors = lines(synthetic{s}.k);
    hold on
    for c = 1:synthetic{s}.k
        plot(x(synthetic{s}.id==c,1),x(synthetic{s}.id==c,2),...
            '.','MarkerSize',20,'Color',colors(c,:))
    end
    hold off
    title('Points and clusters')
    drawnow
    
    % Plot FKM solution
    subplot(2,2,2)
    points_FKM(x,e_fkm)
    title('FKM')
    drawnow
    
    % Plot FMMdd solution
    subplot(2,2,3)
    points_FMMdd(x,e_fmmdd,v_fmmdd)
    title('FMMdd')
    drawnow
    
    % Plot CFKM solution
    subplot(2,2,4)
    points_CFKM(x,e_cfkm)
    title('CFKM')
    drawnow
end

File: python/example2.py
from matplotlib.pyplot import show,subplot,plot,title
import numpy as np
from sklearn.decomposition import PCA
import fuzzykmedoids as fkm
import plots

## Load data
from synthetic import data

for sim in data:
	## Set the number of clusters and fuzziness and compute the distance matrix
	k = sim['k']
	h = 1.5
	g = 1.5
	d = [[sum([(i[s]-j[s])**2 for s in range(len(i))]) for j in sim['x']] for i in sim['x']]
	
	## Solve FKM, FMMdd, CFKM
	Z_fkm,e_fkm = fkm.FKM(d,k,h)
	Z_fmmdd,e_fmmdd,v_fmmdd = fkm.FMMdd(d,k,h)
	Z_cfkm,e_cfkm = fkm.CFKM(d,k,h)
	
	## Plot FKM, FMMdd and CFKM solutions
	# Compute the PCA-projected points in two dimensions
	if len(sim['x'][0]) == 2:
		x = np.array(sim['x'])
	else:
		pca = PCA(n_components=2)
		x = pca.fit_transform(np.array(sim['x']))
	idx = np.array(sim['id'])
	
	# Plot actual points and clusters
	subplot(221)
	for i in range(sim['k']):
		plot(x[idx==i,0],x[idx==i,1],'.')
	title('Points and clusters')
	
	# Plot FKM solution
	subplot(222)
	plots.FKM_points(x,e_fkm)
	title('FKM')
	
	# Plot FMMdd solution
	subplot(223)
	plots.FMMdd_points(x,e_fmmdd,v_fmmdd)
	title('FMMdd')
	
	# Plot CFKM solution
	subplot(224)
	plots.CFKM_points(x,e_cfkm)
	title('CFKM')
	show()

